File: playground/reflected_pipeline.py
"""Reflection-driven pipeline setup for wgpu.

Reads .lux.json sidecar files and creates wgpu bind group layouts,
pipeline layouts, and vertex buffer layouts automatically — no hardcoded
descriptor sets or vertex formats needed.

Usage:
    import json
    with open("shader.frag.json") as f:
        reflection = json.load(f)
    pipeline = ReflectedPipeline(device, vert_reflection, frag_reflection,
                                 vert_module, frag_module, render_format)
    bind_groups = pipeline.create_bind_groups(device, {
        "MVP": mvp_buffer,
        "Light": light_buffer,
        "albedo_tex": (sampler, texture_view),
    })
"""

from __future__ import annotations

import json
from pathlib import Path

import wgpu


# Map from reflection format strings to wgpu VertexFormat
_FORMAT_MAP = {
    "R32_SFLOAT": wgpu.VertexFormat.float32,
    "R32G32_SFLOAT": wgpu.VertexFormat.float32x2,
    "R32G32B32_SFLOAT": wgpu.VertexFormat.float32x3,
    "R32G32B32A32_SFLOAT": wgpu.VertexFormat.float32x4,
    "R32_SINT": wgpu.VertexFormat.sint32,
    "R32G32_SINT": wgpu.VertexFormat.sint32x2,
    "R32G32B32_SINT": wgpu.VertexFormat.sint32x3,
    "R32G32B32A32_SINT": wgpu.VertexFormat.sint32x4,
    "R32_UINT": wgpu.VertexFormat.uint32,
    "R32G32_UINT": wgpu.VertexFormat.uint32x2,
    "R32G32B32_UINT": wgpu.VertexFormat.uint32x3,
    "R32G32B32A32_UINT": wgpu.VertexFormat.uint32x4,
}


def load_reflection(json_path: Path) -> dict:
    """Load a .lux.json reflection file."""
    return json.loads(json_path.read_text(encoding="utf-8"))


def create_default_texture(device: wgpu.GPUDevice):
    """Create a 1x1 white RGBA texture for missing sampler bindings."""
    import numpy as np
    data = np.array([255, 255, 255, 255], dtype=np.uint8)
    texture = device.create_texture(
        size=(1, 1, 1),
        format=wgpu.TextureFormat.rgba8unorm,
        usage=wgpu.TextureUsage.TEXTURE_BINDING | wgpu.TextureUsage.COPY_DST,
    )
    device.queue.write_texture(
        wgpu.TexelCopyTextureInfo(texture=texture),
        data.tobytes(),
        wgpu.TexelCopyBufferLayout(offset=0, bytes_per_row=4, rows_per_image=1),
        (1, 1, 1),
    )
    view = texture.create_view()
    sampler = device.create_sampler(
        mag_filter=wgpu.FilterMode.linear,
        min_filter=wgpu.FilterMode.linear,
    )
    return sampler, view


class ReflectedPipeline:
    """Data-driven pipeline created from reflection metadata.

    Merges vertex and fragment reflection data to create a complete
    pipeline with proper bind group layouts and vertex buffer layout.
    """

    def __init__(
        self,
        device: wgpu.GPUDevice,
        vert_reflection: dict,
        frag_reflection: dict,
        vert_module: wgpu.GPUShaderModule,
        frag_module: wgpu.GPUShaderModule,
        color_format: wgpu.TextureFormat = wgpu.TextureFormat.rgba8unorm,
        depth_format: wgpu.TextureFormat = wgpu.TextureFormat.depth24plus,
        cull_mode: wgpu.CullMode = wgpu.CullMode.back,
    ):
        self.vert_reflection = vert_reflection
        self.frag_reflection = frag_reflection

        # Merge descriptor sets from both stages
        merged_sets = self._merge_descriptor_sets()

        # Create bind group layouts
        self.bind_group_layouts: dict[int, wgpu.GPUBindGroupLayout] = {}
        self._binding_info: dict[int, list[dict]] = {}  # set -> list of binding info

        for set_idx in sorted(merged_sets.keys()):
            entries = []
            binding_info = []
            for binding_data in sorted(merged_sets[set_idx], key=lambda b: b["binding"]):
                entry = self._create_bind_group_layout_entry(binding_data)
                entries.append(entry)
                binding_info.append(binding_data)

            self.bind_group_layouts[set_idx] = device.create_bind_group_layout(
                entries=entries
            )
            self._binding_info[set_idx] = binding_info

        # Create pipeline layout (sorted by set index)
        max_set = max(self.bind_group_layouts.keys()) if self.bind_group_layouts else -1
        layout_list = []
        for i in range(max_set + 1):
            if i in self.bind_group_layouts:
                layout_list.append(self.bind_group_layouts[i])
            else:
                # Create empty layout for gaps
                layout_list.append(device.create_bind_group_layout(entries=[]))

        self.pipeline_layout = device.create_pipeline_layout(
            bind_group_layouts=layout_list
        )

        # Create vertex buffer layout from vertex reflection
        vertex_buffers = self._create_vertex_buffers()

        # Create render pipeline
        self.pipeline = device.create_render_pipeline(
            layout=self.pipeline_layout,
            vertex=wgpu.VertexState(
                module=vert_module,
                entry_point="main",
                buffers=vertex_buffers,
            ),
            primitive=wgpu.PrimitiveState(
                topology=wgpu.PrimitiveTopology.triangle_list,
                cull_mode=cull_mode,
                front_face=wgpu.FrontFace.cw,
            ),
            depth_stencil=wgpu.DepthStencilState(
                format=depth_format,
                depth_write_enabled=True,
                depth_compare=wgpu.CompareFunction.less,
            ),
            multisample=wgpu.MultisampleState(count=1, mask=0xFFFFFFFF),
            fragment=wgpu.FragmentState(
                module=frag_module,
                entry_point="main",
                targets=[
                    wgpu.ColorTargetState(format=color_format),
                ],
            ),
        )

    def create_bind_groups(
        self,
        device: wgpu.GPUDevice,
        resources: dict,
    ) -> dict[int, wgpu.GPUBindGroup]:
        """Create bind groups from a name->resource mapping.

        Resources can be:
        - wgpu.GPUBuffer: for uniform_buffer bindings (auto-wraps in BufferBinding)
        - (sampler, texture_view) tuple: for sampler+sampled_image pairs
        - wgpu.GPUSampler: for standalone sampler bindings
        - wgpu.GPUTextureView: for standalone texture bindings

        The mapping is by name as it appears in the reflection JSON.
        """
        bind_groups = {}

        for set_idx, layout in self.bind_group_layouts.items():
            entries = []
            for binding_data in self._binding_info[set_idx]:
                name = binding_data["name"]
                binding_num = binding_data["binding"]
                btype = binding_data["type"]

                if name not in resources:
                    continue

                resource = resources[name]

                if btype == "uniform_buffer":
                    buf = resource
                    size = binding_data.get("size", 0)
                    entries.append(wgpu.BindGroupEntry(
                        binding=binding_num,
                        resource=wgpu.BufferBinding(buffer=buf, size=size),
                    ))

                elif btype == "sampler":
                    # Expect (sampler, texture_view) tuple
                    if isinstance(resource, tuple):
                        sampler, _ = resource
                    else:
                        sampler = resource
                    entries.append(wgpu.BindGroupEntry(
                        binding=binding_num,
                        resource=sampler,
                    ))

                elif btype in ("sampled_image", "sampled_cube_image"):
                    # Expect (sampler, texture_view) tuple
                    if isinstance(resource, tuple):
                        _, tex_view = resource
                    else:
                        tex_view = resource
                    entries.append(wgpu.BindGroupEntry(
                        binding=binding_num,
                        resource=tex_view,
                    ))

                elif btype == "storage_image":
                    entries.append(wgpu.BindGroupEntry(
                        binding=binding_num,
                        resource=resource,
                    ))

            if entries:
                bind_groups[set_idx] = device.create_bind_group(
                    layout=layout,
                    entries=entries,
                )

        return bind_groups

    def _merge_descriptor_sets(self) -> dict[int, list[dict]]:
        """Merge descriptor sets from vertex and fragment reflections."""
        merged: dict[int, dict[int, dict]] = {}  # set -> {binding -> data}

        for reflection in [self.vert_reflection, self.frag_reflection]:
            for set_str, bindings in reflection.get("descriptor_sets", {}).items():
                set_idx = int(set_str)
                if set_idx not in merged:
                    merged[set_idx] = {}
                for b in bindings:
                    binding_num = b["binding"]
                    if binding_num in merged[set_idx]:
                        # Merge stage flags
                        existing = merged[set_idx][binding_num]
                        existing_flags = set(existing.get("stage_flags", []))
                        new_flags = set(b.get("stage_flags", []))
                        existing["stage_flags"] = list(existing_flags | new_flags)
                    else:
                        merged[set_idx][binding_num] = dict(b)

        return {k: list(v.values()) for k, v in merged.items()}

    def _create_bind_group_layout_entry(self, binding_data: dict) -> wgpu.BindGroupLayoutEntry:
        """Create a single bind group layout entry from reflection data."""
        binding_num = binding_data["binding"]
        btype = binding_data["type"]

        # Determine visibility
        stage_flags = binding_data.get("stage_flags", [])
        if not stage_flags:
            # Default: both stages
            visibility = wgpu.ShaderStage.VERTEX | wgpu.ShaderStage.FRAGMENT
        else:
            visibility = 0
            if "vertex" in stage_flags:
                visibility |= wgpu.ShaderStage.VERTEX
            if "fragment" in stage_flags:
                visibility |= wgpu.ShaderStage.FRAGMENT
            if visibility == 0:
                visibility = wgpu.ShaderStage.VERTEX | wgpu.ShaderStage.FRAGMENT

        if btype == "uniform_buffer":
            return wgpu.BindGroupLayoutEntry(
                binding=binding_num,
                visibility=visibility,
                buffer=wgpu.BufferBindingLayout(type=wgpu.BufferBindingType.uniform),
            )
        elif btype == "sampler":
            return wgpu.BindGroupLayoutEntry(
                binding=binding_num,
                visibility=visibility,
                sampler=wgpu.SamplerBindingLayout(type=wgpu.SamplerBindingType.filtering),
            )
        elif btype == "sampled_image":
            return wgpu.BindGroupLayoutEntry(
                binding=binding_num,
                visibility=visibility,
                texture=wgpu.TextureBindingLayout(
                    sample_type=wgpu.TextureSampleType.float,
                    view_dimension=wgpu.TextureViewDimension.d2,
                ),
            )
        elif btype == "sampled_cube_image":
            return wgpu.BindGroupLayoutEntry(
                binding=binding_num,
                visibility=visibility,
                texture=wgpu.TextureBindingLayout(
                    sample_type=wgpu.TextureSampleType.float,
                    view_dimension=wgpu.TextureViewDimension.cube,
                ),
            )
        elif btype == "storage_image":
            return wgpu.BindGroupLayoutEntry(
                binding=binding_num,
                visibility=visibility,
                storage_texture=wgpu.StorageTextureBindingLayout(
                    access=wgpu.StorageTextureAccess.write_only,
                    format=wgpu.TextureFormat.rgba8unorm,
                    view_dimension=wgpu.TextureViewDimension.d2,
                ),
            )
        elif btype == "acceleration_structure":
            # wgpu doesn't directly support RT — placeholder
            return wgpu.BindGroupLayoutEntry(
                binding=binding_num,
                visibility=visibility,
                buffer=wgpu.BufferBindingLayout(type=wgpu.BufferBindingType.read_only_storage),
            )
        else:
            raise ValueError(f"Unknown binding type in reflection: {btype}")

    def _create_vertex_buffers(self) -> list[wgpu.VertexBufferLayout]:
        """Create vertex buffer layout from vertex reflection."""
        attrs = self.vert_reflection.get("vertex_attributes", [])
        stride = self.vert_reflection.get("vertex_stride", 0)

        if not attrs:
            return []

        wgpu_attrs = []
        for attr in attrs:
            fmt_str = attr.get("format", "R32G32B32A32_SFLOAT")
            fmt = _FORMAT_MAP.get(fmt_str, wgpu.VertexFormat.float32x4)
            wgpu_attrs.append(wgpu.VertexAttribute(
                format=fmt,
                offset=attr["offset"],
                shader_location=attr["location"],
            ))

        return [wgpu.VertexBufferLayout(
            array_stride=stride,
            step_mode=wgpu.VertexStepMode.vertex,
            attributes=wgpu_attrs,
        )]

    @classmethod
    def create_fullscreen(
        cls,
        device: wgpu.GPUDevice,
        frag_reflection: dict,
        frag_module: wgpu.GPUShaderModule,
        vert_module: wgpu.GPUShaderModule,
        color_format: wgpu.TextureFormat = wgpu.TextureFormat.rgba8unorm,
    ) -> "ReflectedPipeline":
        """Create a fullscreen pipeline (no vertex buffers, no depth test)."""
        instance = cls.__new__(cls)
        instance.vert_reflection = {}
        instance.frag_reflection = frag_reflection

        # Only fragment descriptor sets
        merged_sets = {}
        for set_str, bindings in frag_reflection.get("descriptor_sets", {}).items():
            set_idx = int(set_str)
            merged_sets[set_idx] = {b["binding"]: dict(b) for b in bindings}

        instance.bind_group_layouts = {}
        instance._binding_info = {}
        for set_idx in sorted(merged_sets.keys()):
            entries = []
            binding_info = []
            for binding_data in sorted(merged_sets[set_idx].values(), key=lambda b: b["binding"]):
                entry = instance._create_bind_group_layout_entry(binding_data)
                entries.append(entry)
                binding_info.append(binding_data)
            instance.bind_group_layouts[set_idx] = device.create_bind_group_layout(entries=entries)
            instance._binding_info[set_idx] = binding_info

        max_set = max(instance.bind_group_layouts.keys()) if instance.bind_group_layouts else -1
        layout_list = []
        for i in range(max_set + 1):
            if i in instance.bind_group_layouts:
                layout_list.append(instance.bind_group_layouts[i])
            else:
                layout_list.append(device.create_bind_group_layout(entries=[]))

        instance.pipeline_layout = device.create_pipeline_layout(bind_group_layouts=layout_list)

        instance.pipeline = device.create_render_pipeline(
            layout=instance.pipeline_layout,
            vertex=wgpu.VertexState(module=vert_module, entry_point="main", buffers=[]),
            primitive=wgpu.PrimitiveState(topology=wgpu.PrimitiveTopology.triangle_list),
            multisample=wgpu.MultisampleState(count=1, mask=0xFFFFFFFF),
            fragment=wgpu.FragmentState(
                module=frag_module, entry_point="main",
                targets=[wgpu.ColorTargetState(format=color_format)],
            ),
        )
        return instance

    def create_bind_groups_with_defaults(
        self,
        device: wgpu.GPUDevice,
        resources: dict,
    ) -> dict[int, wgpu.GPUBindGroup]:
        """Create bind groups, filling missing resources with defaults."""
        filled = dict(resources)
        default_tex = None

        for set_idx, bindings in self._binding_info.items():
            for b in bindings:
                name = b["name"]
                if name in filled:
                    continue
                if b["type"] in ("sampler", "sampled_image", "sampled_cube_image"):
                    if default_tex is None:
                        default_tex = create_default_texture(device)
                    filled[name] = default_tex
                elif b["type"] == "uniform_buffer":
                    size = b.get("size", 64)
                    filled[name] = device.create_buffer_with_data(
                        data=bytes(size),
                        usage=wgpu.BufferUsage.UNIFORM,
                    )

        return self.create_bind_groups(device, filled)

File: playground_web/src/reflected-pipeline.ts
/**
 * Data-driven pipeline creation from Lux reflection JSON.
 *
 * Port of playground/reflected_pipeline.py — maps reflection metadata
 * to WebGPU bind group layouts, pipeline layouts, and render/compute pipelines.
 */

import type { ReflectionJSON, DescriptorBinding, PushEmulated } from './shader-loader';

// --------------------------------------------------------------------------
// Vertex format mapping (reflection format string → WebGPU)
// --------------------------------------------------------------------------

const FORMAT_MAP: Record<string, GPUVertexFormat> = {
  'R32_SFLOAT': 'float32',
  'R32G32_SFLOAT': 'float32x2',
  'R32G32B32_SFLOAT': 'float32x3',
  'R32G32B32A32_SFLOAT': 'float32x4',
  'R32_SINT': 'sint32',
  'R32G32_SINT': 'sint32x2',
  'R32G32B32_SINT': 'sint32x3',
  'R32G32B32A32_SINT': 'sint32x4',
  'R32_UINT': 'uint32',
  'R32G32_UINT': 'uint32x2',
  'R32G32B32_UINT': 'uint32x3',
  'R32G32B32A32_UINT': 'uint32x4',
};

// --------------------------------------------------------------------------
// Binding type → GPUBindGroupLayoutEntry
// --------------------------------------------------------------------------

function bindingLayoutEntry(
  b: DescriptorBinding,
  binding: number,
  visibility: GPUShaderStageFlags,
): GPUBindGroupLayoutEntry {
  const base: GPUBindGroupLayoutEntry = { binding, visibility };

  switch (b.type) {
    case 'uniform_buffer':
      return { ...base, buffer: { type: 'uniform' } };
    case 'storage_buffer':
      return { ...base, buffer: { type: 'read-only-storage' } };
    case 'sampler':
      return { ...base, sampler: { type: 'filtering' } };
    case 'sampled_image':
      return { ...base, texture: { sampleType: 'float', viewDimension: '2d' } };
    case 'sampled_cube_image':
      return { ...base, texture: { sampleType: 'float', viewDimension: 'cube' } };
    case 'storage_image':
      return { ...base, storageTexture: { access: 'write-only', format: 'rgba8unorm', viewDimension: '2d' } };
    default:
      console.warn(`Unknown binding type: ${b.type}`);
      return { ...base, buffer: { type: 'uniform' } };
  }
}

function stageToVisibility(stageFlags: string[]): GPUShaderStageFlags {
  let vis: GPUShaderStageFlags = 0;
  for (const s of stageFlags) {
    if (s === 'vertex') vis |= GPUShaderStage.VERTEX;
    else if (s === 'fragment') vis |= GPUShaderStage.FRAGMENT;
    else if (s === 'compute') vis |= GPUShaderStage.COMPUTE;
  }
  return vis;
}

// --------------------------------------------------------------------------
// Merge descriptor sets across vertex + fragment reflections
// --------------------------------------------------------------------------

function mergeDescriptorSets(
  vertReflection: ReflectionJSON | null,
  fragReflection: ReflectionJSON | null,
): Map<number, DescriptorBinding[]> {
  const merged = new Map<number, DescriptorBinding[]>();

  function addBindings(refl: ReflectionJSON | null) {
    if (!refl) return;
    for (const [setStr, bindings] of Object.entries(refl.descriptor_sets)) {
      const setNum = parseInt(setStr, 10);
      if (!merged.has(setNum)) merged.set(setNum, []);
      const existing = merged.get(setNum)!;
      for (const b of bindings) {
        // Check for duplicate binding numbers (shared across stages)
        const dup = existing.find(e => e.binding === b.binding);
        if (dup) {
          // Merge stage_flags
          for (const sf of b.stage_flags) {
            if (!dup.stage_flags.includes(sf)) dup.stage_flags.push(sf);
          }
        } else {
          existing.push({ ...b });
        }
      }
    }
  }

  addBindings(vertReflection);
  addBindings(fragReflection);
  return merged;
}

// --------------------------------------------------------------------------
// ReflectedPipeline
// --------------------------------------------------------------------------

export interface PipelineResources {
  /** Resource name → GPUBuffer | GPUSampler | GPUTextureView */
  [name: string]: GPUBuffer | GPUSampler | GPUTextureView;
}

export class ReflectedPipeline {
  readonly renderPipeline: GPURenderPipeline;
  readonly pipelineLayout: GPUPipelineLayout;
  readonly bindGroupLayouts: GPUBindGroupLayout[];
  readonly vertReflection: ReflectionJSON | null;
  readonly fragReflection: ReflectionJSON | null;
  readonly pushEmulated: PushEmulated | null;
  private readonly _setBindingMap: Map<number, DescriptorBinding[]>;

  private constructor(
    renderPipeline: GPURenderPipeline,
    pipelineLayout: GPUPipelineLayout,
    bindGroupLayouts: GPUBindGroupLayout[],
    vertReflection: ReflectionJSON | null,
    fragReflection: ReflectionJSON | null,
    pushEmulated: PushEmulated | null,
    setBindingMap: Map<number, DescriptorBinding[]>,
  ) {
    this.renderPipeline = renderPipeline;
    this.pipelineLayout = pipelineLayout;
    this.bindGroupLayouts = bindGroupLayouts;
    this.vertReflection = vertReflection;
    this.fragReflection = fragReflection;
    this.pushEmulated = pushEmulated;
    this._setBindingMap = setBindingMap;
  }

  /**
   * Create a render pipeline from vertex + fragment reflection metadata.
   */
  static create(
    device: GPUDevice,
    vertModule: GPUShaderModule,
    fragModule: GPUShaderModule,
    vertReflection: ReflectionJSON,
    fragReflection: ReflectionJSON,
    colorFormat: GPUTextureFormat,
    depthFormat: GPUTextureFormat = 'depth24plus',
    cullMode: GPUCullMode = 'back',
    strideOverride = 0,
  ): ReflectedPipeline {
    const setBindings = mergeDescriptorSets(vertReflection, fragReflection);

    // Merge push_emulated from either reflection
    const pushEmulated = vertReflection.push_emulated ?? fragReflection.push_emulated ?? null;

    // Determine max set index (including push emulated set and any gaps)
    let maxSet = -1;
    for (const s of setBindings.keys()) maxSet = Math.max(maxSet, s);
    if (pushEmulated) maxSet = Math.max(maxSet, pushEmulated.set);

    // Create bind group layouts for each set (including empty ones for gaps)
    const bindGroupLayouts: GPUBindGroupLayout[] = [];
    for (let s = 0; s <= maxSet; s++) {
      const bindings = setBindings.get(s);
      if (bindings && bindings.length > 0) {
        const entries = bindings.map(b =>
          bindingLayoutEntry(b, b.binding, stageToVisibility(b.stage_flags)),
        );
        bindGroupLayouts.push(device.createBindGroupLayout({ entries, label: `set${s}` }));
      } else if (pushEmulated && s === pushEmulated.set) {
        // Push constant emulation uniform
        bindGroupLayouts.push(device.createBindGroupLayout({
          entries: [{
            binding: 0,
            visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
            buffer: { type: 'uniform' },
          }],
          label: `set${s}_push_emulated`,
        }));
      } else {
        // Empty layout for gap
        bindGroupLayouts.push(device.createBindGroupLayout({ entries: [], label: `set${s}_empty` }));
      }
    }

    const pipelineLayout = device.createPipelineLayout({
      bindGroupLayouts,
      label: 'reflected_pipeline_layout',
    });

    // Vertex buffer layout from reflection
    const vertexBuffers: GPUVertexBufferLayout[] = [];
    if (vertReflection.vertex_attributes.length > 0) {
      const attrs: GPUVertexAttribute[] = vertReflection.vertex_attributes.map(a => ({
        shaderLocation: a.location,
        offset: a.offset,
        format: FORMAT_MAP[a.format] ?? 'float32x4',
      }));
      vertexBuffers.push({
        arrayStride: strideOverride || vertReflection.vertex_stride,
        stepMode: 'vertex',
        attributes: attrs,
      });
    }

    const renderPipeline = device.createRenderPipeline({
      layout: pipelineLayout,
      vertex: {
        module: vertModule,
        entryPoint: 'main',
        buffers: vertexBuffers,
      },
      fragment: {
        module: fragModule,
        entryPoint: 'main',
        targets: [{ format: colorFormat }],
      },
      primitive: {
        topology: 'triangle-list',
        cullMode,
        frontFace: 'ccw',
      },
      depthStencil: {
        format: depthFormat,
        depthWriteEnabled: true,
        depthCompare: 'less',
      },
      label: 'reflected_render_pipeline',
    });

    return new ReflectedPipeline(
      renderPipeline, pipelineLayout, bindGroupLayouts,
      vertReflection, fragReflection, pushEmulated, setBindings,
    );
  }

  /**
   * Create a fullscreen pipeline (no vertex buffers, no depth test).
   */
  static createFullscreen(
    device: GPUDevice,
    vertModule: GPUShaderModule,
    fragModule: GPUShaderModule,
    fragReflection: ReflectionJSON,
    colorFormat: GPUTextureFormat,
  ): ReflectedPipeline {
    const setBindings = mergeDescriptorSets(null, fragReflection);
    const pushEmulated = fragReflection.push_emulated ?? null;

    let maxSet = -1;
    for (const s of setBindings.keys()) maxSet = Math.max(maxSet, s);
    if (pushEmulated) maxSet = Math.max(maxSet, pushEmulated.set);

    const bindGroupLayouts: GPUBindGroupLayout[] = [];
    for (let s = 0; s <= maxSet; s++) {
      const bindings = setBindings.get(s);
      if (bindings && bindings.length > 0) {
        const entries = bindings.map(b =>
          bindingLayoutEntry(b, b.binding, stageToVisibility(b.stage_flags)),
        );
        bindGroupLayouts.push(device.createBindGroupLayout({ entries, label: `set${s}` }));
      } else if (pushEmulated && s === pushEmulated.set) {
        bindGroupLayouts.push(device.createBindGroupLayout({
          entries: [{
            binding: 0,
            visibility: GPUShaderStage.FRAGMENT,
            buffer: { type: 'uniform' },
          }],
          label: `set${s}_push_emulated`,
        }));
      } else {
        bindGroupLayouts.push(device.createBindGroupLayout({ entries: [], label: `set${s}_empty` }));
      }
    }

    const pipelineLayout = device.createPipelineLayout({ bindGroupLayouts });

    const renderPipeline = device.createRenderPipeline({
      layout: pipelineLayout,
      vertex: { module: vertModule, entryPoint: 'main' },
      fragment: {
        module: fragModule,
        entryPoint: 'main',
        targets: [{ format: colorFormat }],
      },
      primitive: { topology: 'triangle-list' },
      label: 'fullscreen_pipeline',
    });

    return new ReflectedPipeline(
      renderPipeline, pipelineLayout, bindGroupLayouts,
      null, fragReflection, pushEmulated, setBindings,
    );
  }

  /**
   * Create bind groups from a name → resource mapping.
   */
  createBindGroups(
    device: GPUDevice,
    resources: PipelineResources,
  ): GPUBindGroup[] {
    const groups: GPUBindGroup[] = [];

    for (let s = 0; s < this.bindGroupLayouts.length; s++) {
      // Push emulated bind group
      if (this.pushEmulated && s === this.pushEmulated.set) {
        const buf = resources['_push_emulated'] as GPUBuffer;
        if (buf) {
          groups.push(device.createBindGroup({
            layout: this.bindGroupLayouts[s],
            entries: [{ binding: 0, resource: { buffer: buf } }],
          }));
          continue;
        }
      }

      const bindings = this._setBindingMap.get(s);
      if (!bindings || bindings.length === 0) {
        groups.push(device.createBindGroup({
          layout: this.bindGroupLayouts[s],
          entries: [],
        }));
        continue;
      }

      const entries: GPUBindGroupEntry[] = [];
      for (const b of bindings) {
        const res = resources[b.name];
        if (!res) {
          console.warn(`Missing resource for binding '${b.name}' (set=${s}, binding=${b.binding})`);
          continue;
        }

        if (b.type === 'uniform_buffer' || b.type === 'storage_buffer') {
          entries.push({ binding: b.binding, resource: { buffer: res as GPUBuffer } });
        } else if (b.type === 'sampler') {
          entries.push({ binding: b.binding, resource: res as GPUSampler });
        } else if (b.type === 'sampled_image' || b.type === 'sampled_cube_image') {
          entries.push({ binding: b.binding, resource: res as GPUTextureView });
        } else if (b.type === 'storage_image') {
          entries.push({ binding: b.binding, resource: res as GPUTextureView });
        }
      }

      groups.push(device.createBindGroup({
        layout: this.bindGroupLayouts[s],
        entries,
        label: `bind_group_set${s}`,
      }));
    }

    return groups;
  }
}

// --------------------------------------------------------------------------
// Compute pipeline helper
// --------------------------------------------------------------------------

export function createReflectedComputePipeline(
  device: GPUDevice,
  module: GPUShaderModule,
  reflection: ReflectionJSON,
): { pipeline: GPUComputePipeline; layout: GPUPipelineLayout; bindGroupLayouts: GPUBindGroupLayout[] } {
  const setBindings = mergeDescriptorSets(null, reflection);
  const pushEmulated = reflection.push_emulated ?? null;

  let maxSet = -1;
  for (const s of setBindings.keys()) maxSet = Math.max(maxSet, s);
  if (pushEmulated) maxSet = Math.max(maxSet, pushEmulated.set);

  const bindGroupLayouts: GPUBindGroupLayout[] = [];
  for (let s = 0; s <= maxSet; s++) {
    const bindings = setBindings.get(s);
    if (bindings && bindings.length > 0) {
      const entries = bindings.map(b =>
        bindingLayoutEntry(b, b.binding, GPUShaderStage.COMPUTE),
      );
      bindGroupLayouts.push(device.createBindGroupLayout({ entries }));
    } else if (pushEmulated && s === pushEmulated.set) {
      bindGroupLayouts.push(device.createBindGroupLayout({
        entries: [{
          binding: 0,
          visibility: GPUShaderStage.COMPUTE,
          buffer: { type: 'uniform' },
        }],
      }));
    } else {
      bindGroupLayouts.push(device.createBindGroupLayout({ entries: [] }));
    }
  }

  const layout = device.createPipelineLayout({ bindGroupLayouts });
  const pipeline = device.createComputePipeline({
    layout,
    compute: { module, entryPoint: 'main' },
  });

  return { pipeline, layout, bindGroupLayouts };
}
